File: src/java/magma/api/result/Err.java
package magma.api.result;

import magma.api.Tuple;

import java.util.Optional;
import java.util.function.Function;
import java.util.function.Supplier;

public class Err<T, X> implements Result<T, X> {
    private final X error;

    public Err(X error) {
        this.error = error;
    }

    @Override
    public Optional<T> findValue() {
        return Optional.empty();
    }

    @Override
    public Optional<X> findError() {
        return Optional.of(error);
    }

    @Override
    public <R> Result<R, X> mapValue(Function<T, R> mapper) {
        return new Err<>(error);
    }

    @Override
    public <R> Result<R, X> flatMapValue(Function<T, Result<R, X>> mapper) {
        return new Err<>(error);
    }

    @Override
    public <R> Result<Tuple<T, R>, X> and(Supplier<Result<R, X>> other) {
        return new Err<>(error);
    }

    @Override
    public boolean isOk() {
        return false;
    }

    @Override
    public <R> Result<T, R> mapErr(Function<X, R> mapper) {
        return new Err<>(mapper.apply(error));
    }

    @Override
    public <R> R match(Function<T, R> onOk, Function<X, R> onErr) {
        return onErr.apply(error);
    }
}

File: src/c/magma/api/result/Err.c
import magma.api.Tuple;import magma.api.option.None;import magma.api.option.Option;import magma.api.option.Some;import java.util.Optional;struct Err<T, X>(X error){
	<R>Result<R, X> flatMapValue(Tuple<any*, Result<R, X> (*)(any*, T)> mapper){
		return new Err<>(this.error);
	}
	<R>Result<R, X> mapValue(Tuple<any*, R (*)(any*, T)> mapper){
		return new Err<>(this.error);
	}
	<R>Result<T, R> mapErr(Tuple<any*, R (*)(any*, X)> mapper){
		return new Err<>(mapper.apply(this.error));
	}
	<R>R match(Tuple<any*, R (*)(any*, T)> onOk, Tuple<any*, R (*)(any*, X)> onErr){
		return onErr.apply(this.error);
	}
	<R>Result<Tuple<T, R>, X> and(Tuple<any*, Result<R, X> (*)(any*)> other){
		return new Err<>(this.error);
	}
	<R>Result<T, Tuple<X, R>> or(Tuple<any*, Result<T, R> (*)(any*)> other){
		return other.get().mapErr(()->new Tuple<>(this.error, otherErr));
	}
	boolean isOk(){
		return false;
	}
	Optional<X> findError0(){
		return Optional.of(this.error);
	}
	Option<X> findError(){
		return findError0().<Option<X>>map(Some::new).orElseGet(None::new);
	}
	Result<T, X> Result(){
		return Result.new();
	}
}